### PR TITLE
[Desktop] Add store and new algorithm pages

### LINKED ISSUE #82: Add algorithms store tab

### LINKED ISSUE #83: Add tab for adding new algorithm

### PR DESCRIPTION
**Pull request by AntonDudchenko**
close #82 
close #83

### CODE REVIEW

File: webforces/server/structs.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ def fromDict(cls, dict):
 
 @dataclass
 class Algorithm:
-    alg_id: int
+    alg_id: int = -1
```
Comment (aobolensk) [inline webforces/server/structs.py]: This is done intentionally. You should provide an ID. If you don't have any just use 0.

File: webforces/server/mongodbworker.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ def getAlgByTitle(self, title) -> Tuple[DBStatus, Algorithm]:
             algs_collection = self.db["algs"]
             alg_d = algs_collection.find_one({"title": title})
             if alg_d is None:
-                logger.error("This algorithm does not exist")
+                logger.debug("This algorithm does not exist")
```
Comment (aobolensk) [inline webforces/server/mongodbworker.py]: Why do we need to replace it?

File: webforces/api/views.py [hunk 3]
```diff
@@ -116,15 +118,36 @@ class AlgsView(APIView):
 
     def get(self, request):
         core = Core()
-        status, algs = core.db.getAllAlgs()
+        if request.data['fromTitle'] == '1':
```
Comment (aobolensk) [inline webforces/api/views.py]: What does `fromTitle` param mean? We already have separate view for getting algs using title. See `class AlgViewTitle` in current file.

File: desktop_app/GUI/UI/MainWindow.ui [hunk 4]
```diff
@@ -6,8 +6,8 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>906</width>
```
Comment (aobolensk) [inline desktop_app/GUI/UI/MainWindow.ui]: Do we actually need changes related to size in this file?

File: desktop_app/GUI/Py/MainWindow.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
 class Ui_MainWindow(object):
     def setupUi(self, MainWindow):
         MainWindow.setObjectName("MainWindow")
-        MainWindow.resize(906, 720)
```
Comment (aobolensk) [inline desktop_app/GUI/Py/MainWindow.py]: Do we actually need changes related to size in this file?

File: desktop_app/GUI/Py/MainWindow.py [hunk 6]
```diff
@@ -27,11 +27,13 @@ def setupUi(self, MainWindow):
...
-        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
-        self.horizontalLayout = QtWidgets.QHBoxLayout()
+        self.widget1 = QtWidgets.QWidget(self.widget)
+        self.widget1.setGeometry(QtCore.QRect(10, 10, 871, 39))
+        self.widget1.setObjectName("widget1")
+        self.horizontalLayout = QtWidgets.QHBoxLayout(self.widget1)
+        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
         self.horizontalLayout.setObjectName("horizontalLayout")
-        self.storeButton = QtWidgets.QPushButton(self.widget)
+        self.storeButton = QtWidgets.QPushButton(self.widget1)
```
Comment (aobolensk) [inline desktop_app/GUI/Py/MainWindow.py]: Why widget1?

File: desktop_app/GUI/Py/ui.py [hunk 7]
```diff
@@ -140,6 +142,107 @@ def setDisabled(self):
...
+        self.addButton.clicked.connect(self.add)
+
+    def add(self):
+        title = self.nameEdit.text()
+        desc = self.descEdit.text()
+        langText = self.langBox.currentText()
+        langId = -1
+        if langText == "C":
+            langId = 0
+        elif langText == "C++":
```
Comment (aobolensk) [inline desktop_app/GUI/Py/ui.py]: lang_id == unknown language. See: `structs.py`

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Comment by AntonDudchenko** [comment]
image
Я отказываюсь не ресайзить MainWindow

File: webforces/api/views.py [hunk 8]
```diff
@@ -116,15 +118,30 @@ class AlgsView(APIView):
...
+        for alg in algs:
+            data.append(alg.__dict__)
+        return Response(data)
 
-        if status != DBStatus.s_ok:
-            raise Exception("Could not get algs")
+    def post(self, request):
+        core = Core()
+        data = request.data
+        print(data['login'])
```
Comment (aobolensk) [inline webforces/api/views.py]: Remove this debug print, please

**Review by aobolensk — APPROVED** [review]

**Review by skonnov — APPROVED** [review]

**Pull request merged**